### ISSUE TITLE
Support for visualising tenors with ellipses?

### ISSUE THREAD

**Issue by jkshenton** [superstar54/beautiful-atoms#97]
Thanks a lot for making this fantastic tool! 

Would it be possible to add in support for visualising tensors using e.g. ellipsoids? This would be very useful for NMR/muSR/EPR researchers!

ASE supports reading in e.g. .magres files that contain magnetic shielding/electric field gradient tensors and these can be visualised in something like MagresView: https://www.ccpnc.ac.uk/magresview/magresview/magres_view.html but it would be nice to be able to make some high-quality images using blender!

Could something similar be done with beautifulatoms?

**Comment by superstar54** [comment]
Could you please share an example merges file and an example image?

**Comment by jkshenton** [comment]
Sure! Here's a screenshot of MagresView showing magnetic shielding tensors on an example system:

![magresview_ss](https://user-images.githubusercontent.com/10869663/170681434-d69cdeff-2238-448e-ae31-4ab24e561ea4.png)

Here's an example .magres file:
https://github.com/CCP-NC/soprano/blob/master/tests/test_data/ethanol.magres

(There are lots more in [this database](https://www.ccpnc.ac.uk/database).)

ASE reads the tensors into named arrays. For example, 
```python
from ase.io import read
atoms = read('ethanol.magres')
ms_array = atoms.get_array('ms')
# ms_array.shape == (natoms, 3, 3)
```

The ellipsoids for each atom can then drawn using the eigenvectors/eigenvalues of that atom's tensor. 

Thanks!

**Comment by superstar54** [comment]
Thanks for the details. Yes, this could be implemented into Beautiful Atoms. 

Could you show how the ellipses (parameters) can be determined by the eigenvectors/eigenvalues of that atom's tensor?

**Comment by jkshenton** [comment]
I'm not sure what is needed on the blender side of things -- sorry! The 3 eigenvectors define the principal axes of the ellipse and eigenvalues determine the extent along each axis. See e.g. this definition and the conversion to a parametric representation:

https://en.wikipedia.org/wiki/Ellipsoid#As_a_quadric

Does that help?

**Comment by superstar54** [comment]
Yes, that is very helpful. Thanks. I will implement it in the next few days and keep you updated.

**Comment by superstar54** [comment]
One question. In the example you shared, the three eigenvectors of all tensors are not orthogonal. However, for an ellipsoid, the principal axes should be perpendicular to each other.

**Comment by jkshenton** [comment]
Oh, you're right -- sorry about that! It looks like it's the eigenvectors of the symmetric part of the tensor that are needed for this and that the final eigenvector is anyway explicitly computed as the cross product of the first two (at least in the codes I have seen). I had a look in the soprano library that does this sort of analysis and found this:

```python
 self._symm = (self._data + self._data.T) / 2.0
 evals, evecs = np.linalg.eigh(self._symm)
```
(where `self._data` is a 3x3 array with the MS/EFG tensor for a particular atom).
from here: https://github.com/CCP-NC/soprano/blob/991a925b72bc7c3a38cc60c87cded7568a67d906/soprano/nmr/tensor.py#L71

There's also a comment [here](https://github.com/CCP-NC/soprano/blob/991a925b72bc7c3a38cc60c87cded7568a67d906/soprano/nmr/tensor.py#L107) suggesting that after sorting the eigenvectors according to one convention or another, the
`# Last eigenvector must be the cross product of the first two`

So maybe orthogonality is enforced at that point...?

The same thing is done here:
https://github.com/stur86/crystvis-js/blob/master/lib/tensor.js

which is the basis for visualising these tensor ellipsoids in MagresView2: https://stur86.github.io/magresview-2/

**Comment by superstar54** [comment]
Thanks for the information. As shown in #103 , the ellipsoids are correct now. 

However, there is a special issue regarding reading the tensor data into Batoms. We are working on it. Will keep you updated when finished.
